pytket/pytket/mbqc_compile/mpattern.py

```diff
@@ -0,0 +1,434 @@
+        
+        :param c:       A pytket circuit.
+        :param type:    Circuit
+        
+        :returns:       A pyzx circuit.
+        :rtype:         pyzxCircuit
+        """
+        Transform.RebaseToPyZX().apply(c)
+        return tk_to_pyzx(c)
+    
+    def zx_diagram(self, pyzxc: pyzxCircuit) -> Graph:
+        """
+        Converts a pyzx circuit to a zx diagram.
+        
+        :param pyzxc:   A pyzx circuit.
+        :param type:    pyzxCircuit
+        
+        :returns:       A zx diagram.
+        :rtype:         Graph
+        """
+        g = circuit_to_graph(pyzxc)
+        interior_clifford_simp(g, quiet=True)
+        for q in range(self.qubits):
+            self.inputs[q] = sorted(list(g.vertices()))[q]
+            self.outputs[q] = sorted(list(g.vertices()))[-self.qubits+q]
+        self.remove_redundant(g)
+        #We assume that g.copy() will squash the vertex
+        #labels and thus we keep track of the new input/output vertices. If
+        #pyzx is updated such that graph.copy() no longer changes vertex labels
+        #then comment out the next line (label_squish(g)).
+        self.label_squish(g)
+        return g.copy()
+    
+    def label_squish(self, g: Graph) -> None:
+        """
+        Updates the input/output labels of the MPattern to matched a squished
+        graph cause by g.copy().
+        
+        :param g:       A pyzx graph.
+        :param type:    Graph
+        """
+        original_labels = sorted(list(g.vertices()))
+        for i in self.inputs.keys():
+            for v in range(len(original_labels)):
+                if self.inputs[i] == original_labels[v]:
+                    self.inputs[i] = v
+                    break
+        for o in self.outputs.keys():
+            for v in range(len(original_labels)):
+                if self.outputs[o] == original_labels[v]:
+                    self.outputs[o] = v
+                    break
+    
+    def entangle(g: Graph) -> Circuit:
+        """
+        Creates a tket circuit which implements the edges of a zx diagram
+        via CZ gates on pairs of qubits.
+        
+        :param g:       A zx diagram.
+        :param type:    Graph
+        
+        :returns:       A pytket circuit.
+        :rtype:         Circuit
+        """
+        c = Circuit(len(g.vertices()),len(g.vertices()))
+        vlist = list(g.vertices())
+        dlist = []
+        for v in vlist:
+            dlist.append(len(g.neighbors(v)))
+        vlist = [x for _, x in sorted(zip(dlist, vlist))]
+        vlist.reverse()
+        edge_pool = set(g.edge_set())
+        finished_edge_pool = {}
+        doneround = {}
+        for v in vlist:
+            doneround[v] = False
+        while len(edge_pool)>0:
+            for vid in range(len(vlist)):
+                if not doneround[vlist[vid]]:
+                    for vid2 in range(vid+1,len(vlist)):
+                        if ((not doneround[vlist[vid]]) and (not doneround[vlist[vid2]])):
```

Comment:
do we need to check `doneround[vlist[vid]` again?

Reply:
Fixed.